backend/liveness/liveness.go

```diff
@@ -0,0 +1,299 @@
+// for entities. On life and death events, callback functions that are
+// registered on NewSwitchSet are started as new goroutines.
+//
+// The SwitchSet uses the Alive method to both register members of the set,
+// and to assert their liveness once registered. After its first call to
+// Alive, if a member does not assert its liveness, then it will be presumed
+// to be dead, and the callback for dead members will be called.
+//
+// When an entity in a SwitchSet dies, it gains a new life in the underworld.
+// In the underworld, a dead callback is issued for every TTL interval.
+// Entities can go from being dead to alive by calling Alive. When that
+// happens, an entity that lives in the underworld will be reborn.
+type SwitchSet struct {
+	client      *clientv3.Client
+	prefix      string
+	notifyDead  EventFunc
+	notifyAlive EventFunc
+	logger      logrus.FieldLogger
+
+	// This channel serializes events so that their execution ordering is
+	// as expected, without causing undue blocking in the main monitoring
+	// loop.
+	events chan func()
+}
+
+// EventFunc is a function that can be used by a SwitchSet to handle events.
+// The previous state of the switch will be passed to the function.
+//
+// For "dead" EventFuncs, the leader flag can be used to determine if the
+// client that flipped the switch is our client. For "alive" EventFuncs,
+// this parameter is always false.
+type EventFunc func(key string, prev State, leader bool)
+
+// NewSwitchSet creates a new SwitchSet. It will use an etcd prefix of
+// path.Join(SwitchPrefix, name). The dead and live callbacks will be called
+// on all life and death events.
+func NewSwitchSet(client *clientv3.Client, name string, dead, alive EventFunc, logger logrus.FieldLogger) *SwitchSet {
+	return &SwitchSet{
+		client:      client,
+		prefix:      path.Join(SwitchPrefix, name),
+		notifyDead:  dead,
+		notifyAlive: alive,
+		logger:      logger,
+		events:      make(chan func(), 512),
+	}
+}
+
+func (t *SwitchSet) ping(ctx context.Context, key string, ttl int64, alive bool) error {
+	if ttl < FallbackTTL {
+		return fmt.Errorf("bad ttl: %d is less than the minimum value of %d", ttl, FallbackTTL)
+	}
+	putVal := ttl
+	if !alive {
+		putVal = -putVal
+	}
+	key = path.Join(t.prefix, key)
+	val := fmt.Sprintf("%x", putVal)
+	lease, err := t.client.Grant(ctx, ttl)
+	if err != nil {
+		return err
+	}
+	_, err = t.client.Put(ctx, key, val, clientv3.WithLease(lease.ID), clientv3.WithPrevKV())
+	return err
+}
+
+// Alive is an assertion that an entity is alive.
+//
+// If the SwitchSet doesn't know about the entity yet, then it will be
+// registered, and the TTL countdown will start. Unless the entity continually
+// asserts its liveness with calls to Alive, it will be presumed dead.
+//
+// The ttl parameter is the time-to-live in seconds for the entity. The minimum
+// TTL value is 5. If a smaller value is passed, then an error will be returned
```

Comment:
Does this imply that check ttl's must be at least 5 seconds? If so, should we add validation for that?

Reply:
That's right - we will need to add validation for those values when integrating this work in a future PR.

Reply:
Makes sense 👍